[BUGFIX beta] fix for lastObject/firstObject update issue

bekzod opened the pull request:
enhanced #15110 PR, @stefanpenner

bekzod committed:
[BUGFIX beta] fix for lastObject/firstObject update issue

stefanpenner commented on packages/ember-runtime/lib/mixins/array.js:
```diff
@@ -141,19 +141,24 @@ export function arrayContentDidChange(array, startIdx, removeAmt, addAmt) {
-      propertyWillChange(array, 'lastObject', meta);
-      propertyDidChange(array, 'lastObject', meta);
+  if (cache !== undefined) {
+    if (cache.firstObject !== undefined && startIdx === 0 && objectAt(array, 0) !== cacheFor.get(cache, 'firstObject')) {
```
i think one of the goals @hjdivad had, was to avoid the `objectAt(x)` lookups, as they would force the CP to be accessed.

I'll ping @hjdivad   to chime in.

bekzod replied:
yeah but you need to invalidate cp when it changes no ?

stefanpenner replied:
If the index for firstObject / lastObject changed, then we should broadcast. Even if the object is equal. That means we wont need both the objectAt stuff and the range based detection.

bekzod commented:
I updated my PR, so it does not call `objectAt` but now behavior is changed of `firstObject/lastObject` change event, so now it notifies change even if same object as `firstObject/lastObject` is shifted/pushed to `array` , I think it is still better than how was before.
adjusted/added test accordingly @stefanpenner

stefanpenner commented:
@bekzod ya, that was the approach @hjdivad was taking. It is likely a trade-off we are good with.

hjdivad commented on packages/ember-runtime/lib/mixins/array.js:
```diff
@@ -141,19 +141,27 @@ export function arrayContentDidChange(array, startIdx, removeAmt, addAmt) {
+
+      let lastIndex = length - delta - 1;
+      let lastAffectedIndex = startIdx + removedAmount;
+      if (lastIndex < lastAffectedIndex) {
```
I think this will miss cases where `lastObject` changes from negative indices, eg

```js
[0, 1, 2, 3].replace(-2, 2);
// => [0, 1]
```

bekzod replied:
thanks, was not sure if negative startIndex could be passed will handle that

hjdivad replied:
awesome thanks @bekzod 👍

hjdivad commented on packages/ember-runtime/tests/suites/mutable_array/pushObject.js:
```diff
@@ -54,4 +54,26 @@ suite.test('[A,B,C].pushObject(X) => [A,B,C,X] + notify', function() {
   equal(observer.validate('firstObject'), false, 'should NOT have notified firstObject');
 });
 
+suite.test('[A,B,C,C].pushObject(A) => [A,B,C,C] + notify', function() {
```
👍 

We'll need a test for removing the tail from a negative index

bekzod replied:
added that, you could take another shot when you have time, thanks

hjdivad commented on packages/ember-runtime/lib/mixins/array.js:
```diff
@@ -141,19 +141,27 @@ export function arrayContentDidChange(array, startIdx, removeAmt, addAmt) {
-    }
+  if (cache !== undefined) {
+    if (cache.firstObject !== undefined && startIdx === 0) {
+      propertyWillChange(array, 'firstObject');
```
(which is fine, just want to confirm)

hjdivad commented on packages/ember-runtime/tests/suites/mutable_array/replace.js:
```diff
@@ -126,6 +126,27 @@ suite.test('[A,B,C,D].replace(2,2) => [A,B] + notify', function() {
   equal(observer.validate('firstObject'), false, 'should NOT have notified firstObject once');
 });
 
+suite.test('[A,B,C,D].replace(-1,1) => [A,B,C] + notify', function() {
```
👍

hjdivad commented on packages/ember-runtime/lib/mixins/array.js:
```diff
@@ -141,19 +141,29 @@ export function arrayContentDidChange(array, startIdx, removeAmt, addAmt) {
+    let delta = addedAmount - removedAmount;
+    let previousLength = length - delta;
+
+    let normalStartIdx = startIdx < 0 ? previousLength + startIdx : startIdx;
```
👍

hjdivad reviewed (approved):
LGTM

stefanpenner commented:
@krisselden bugfix beta or let it ride the canary train?

stefanpenner commented:
chatted with @krisselden hes ok with `beta`

stefanpenner commented:
@bekzod can you squash, and prefix with `[BUGFIX beta]` ?

bekzod commented:
@stefanpenner done
